gptcache/manager/vector_data/pgvector.py

```diff
@@ -0,0 +1,173 @@
+        "cosine": { "operator": "<=>", "name": "vector_cosine_ops" },
+        "inner_product": { "operator": "<->", "name": "vector_ip_ops" },
+    }
+
+
+    def __init__(
+        self,
+        url: str,
+        index_params: dict,
+        collection_name: str = "gptcache",
+        dimension: int = 0,
+        top_k: int = 1,
+    ):
+        if dimension <= 0:
+            raise ValueError(
+                f"invalid `dim` param: {dimension} in the pgvector store."
+            )
+        self.dimension = dimension
+        self.top_k = top_k
+        self.index_params = index_params
+        self._url = url
+        self._store, self._index = get_model_and_index(
+            collection_name,
+            dimension,
+            index_type = self.INDEX_PARAM[index_params["index_type"]]["name"],
+            lists = index_params["params"]["lists"]
+        )
+        self._connect(url)
+        self._create_collection()
+
+    def _connect(self, url):
+        self._engine = create_engine(url, echo=False)
+        self._session = sessionmaker(bind=self._engine)  # pylint: disable=invalid-name
+
+    def _create_collection(self):
+        with self._engine.connect() as con:
+            con.execution_options(isolation_level="AUTOCOMMIT").execute(text("CREATE EXTENSION IF NOT EXISTS vector;"))
+
+        self._store.__table__.create(bind=self._engine, checkfirst=True)
+        self._index.create(bind=self._engine, checkfirst=True)
+
+    def _query(self, session):
+        return session.query(self._store)
+
+    def _format_data_for_search(self, data):
+        return f"[{','.join(map(str, data))}]"
+
+    def mul_add(self, datas: List[VectorData]):
+        data_array, id_array = map(list, zip(*((data.data, data.id) for data in datas)))
+        np_data = np.array(data_array).astype("float32")
+        entities = [{"id": id, "embedding": embedding.tolist()} for id, embedding in zip(id_array, np_data)]
+
+        with self._session() as session:
+            session.bulk_insert_mappings(self._store, entities)
+            session.commit()
+
+    def search(self, data: np.ndarray, top_k: int = -1):
+        if top_k == -1:
+            top_k = self.top_k
+
+        formatted_data = self._format_data_for_search(data.reshape(1, -1)[0].tolist())
+        index_config = self.INDEX_PARAM[self.index_params["index_type"]]
+        similarity = self._store.embedding.op(index_config["operator"])(formatted_data)
+        with self._session() as session:
+            session.execute(text(f"SET LOCAL ivfflat.probes = {self.index_params['params']['probes'] or 10};"))
+            search_result = self._query(session).add_columns(
+                similarity.label("distances")
+            ).order_by(
+                similarity
+            ).limit(top_k).all()
+
+        return search_result
+
+    def delete(self, ids = None):
+        with self._session() as session:
+            if ids is None:
+                self._query(session).delete()
```

Comment:
the `ids` is never `None`, and the operation is dangerous. So the `if` can be deleted